### ISSUE TITLE
Spurious "Connection died, tried 5 times before giving up" using HTTP/2

### ISSUE THREAD

**Issue by alexcrichton** [curl/curl#7400]
### I did this

Currently the [Cargo package manager for Rust](https://github.com/rust-lang/cargo) uses Curl to download files over HTTP, namely dependency libraries. Recently Cargo, on Rust's nightly channel, updated from curl 7.76.0 to curl 7.77.0. Since updating users are [reporting](https://github.com/rust-lang/crates.io/issues/3782) an increase in errors seen while downloading.

I've personally been seeing this in a number of CI builds that use Cargo with curl 7.77.0, and the errors as-reported by Cargo is:

```
error: failed to download from `https://crates.io/api/v1/crates/wat/1.0.38/download`

Caused by:
  [55] Failed sending data to the peer (Connection died, tried 5 times before giving up)
```

I'm able to reproduce this locally on a machine where it just downloads dependencies in a loop, and frequently it will hit this error while downloading. Since this was an apparent regression in Curl I did a bisection of commits between 7.76.0 and 7.77.0 where on each commit I built that version of Curl into Cargo and then downloaded all of Cargo's own dependencies 100 times, where if any download failed that mean the commit was "bad". Bisection resulted in pointing to this commit -- https://github.com/curl/curl/commit/252790c5335a221179dfbacb40e065750bbe7544 -- where download would fail spuriously after that but would succeed at least 100 times before that.

Unfortunately reproduction is unlikely to be super easy unless you want to dive into Cargo (which I'm assuming you'd rather not). Cargo uses Curl as a library, so Cargo's usage is pretty far removed from the `curl` command line tool for example. In theory Cargo is doing pretty a pretty bland request of "please download all these files over HTTP/2" with some reasonable timeout and retry limits configured. Cargo enables HTTP/2 and pipewait with a maximum of 2 connections per-host at this time.

To hopefully help in assisting with debugging this I captured [debug logs from curl during a failed transfer](https://gist.github.com/alexcrichton/5eb36c4ea3ccc79d28f1a9c16d1202b4). These debug logs were captured at 252790c5335a221179dfbacb40e065750bbe7544 with `DEBUG_HTTP2` enabled as well.

I'm happy to help test out patches and/or get more debugging information if necessary. I can post reproduction steps as well but it requires using Rust to build Cargo and having a nontrivial setup to thread through a custom Curl dependency into Cargo.

### I expected the following

Given the frequency of these errors it seems unlikely that this is caused by an actual spurious error, so the expectation is that something is retried internally as necessary or requeued as necessary without bubbling up an error.

### curl/libcurl version

This relates to curl 7.77.0, specifically an apparent regression introduced with 252790c5335a221179dfbacb40e065750bbe7544

### operating system

The system I tested this on was:

```
Linux 4.19.0-9-arm64 #1 SMP Debian 4.19.118-2+deb10u1 (2020-06-07) aarch64 GNU/Linux
```

but this seems to appear frequently with GitHub Actions runners as well, and I've seen it locally on my macbook too.

**Comment by alexcrichton** [comment]
Some further debugging and I think I've got a way to reproduce this a bit easier. Given [this `urls.txt`](https://gist.github.com/alexcrichton/2192bc5e8c9669155a84a63ef423e695) file I can reproduce the issue with the `curl` command line executable:

```
curl --remote-name-all -Ss $(cat /urls.txt) -L -Z --parallel-max 100 --http2-prior-knowledge
```

I've re-bisected using this command, and bisection again points at https://github.com/curl/curl/commit/252790c5335a221179dfbacb40e065750bbe7544 as the commit between 7.76 and 7.77 where errors started happening.

For reference I used [this `Dockerfile`](https://gist.github.com/alexcrichton/304fdd303ea92210d913d9a3cf869726) for bisection and for nghttp2 it was using https://github.com/nghttp2/nghttp2/commit/31253f400d8055a6dde256a84bfb55dfb788ef6a (current HEAD at the time of me running this). I was also building locally with `./configure --disable-shared --enable-debug --with-openssl --with-nghttp2`.

The custom build also yields:

```
# curl --version --verbose
curl 7.77.0-DEV (aarch64-unknown-linux-gnu) libcurl/7.77.0-DEV OpenSSL/1.1.1f zlib/1.2.11 nghttp2/1.40.0
Release-Date: [unreleased]
Protocols: dict file ftp ftps gopher gophers http https imap imaps mqtt pop3 pop3s rtsp smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS Debug HSTS HTTP2 HTTPS-proxy IPv6 Largefile libz NTLM NTLM_WB SSL TLS-SRP TrackMemory UnixSockets
```

and [these are the verbose logs of a failed transfer](https://gist.github.com/alexcrichton/7b810cede50c32746723d0958eea6b54)

**Comment by alexcrichton** [comment]
Er sorry, I forgot to enable http2 debug logging for those logs. [This is a failure log with `DEBUG_HTTP2` defined in `lib/http2.c`](https://gist.github.com/alexcrichton/9c75520c51c806a8cafd6fc5e484de5a).

Additionally I'm not sure if this is related to https://github.com/curl/curl/issues/5611, it may be the same issue?

**Comment by bagder** [comment]
> curl --remote-name-all -Ss $(cat /urls.txt) -L -

Does this command line reliably reproduce the problem for you? I've run it many times now on my dev box (using current curl built from git) but it always succeeds. Does the problem still remain for you with curl 7.78.0 ?

**Comment by alexcrichton** [comment]
I indeed cannot reproduce on the current master branch, but I believe the reason for that is that nghttp2 is no longer detected after b4b34db65, so http2 isn't used (and there's no bug unless http2 is used).

Inside of `config.log` I see:

```
/usr/bin/ld: cannot find -l-lnghttp2
```

Is this a local configuration error for me? When you build locally do you have http2 enabled with nghttp2?

**Comment by alexcrichton** [comment]
For reference, though, on fbf26594e38ebf6e8b0f0e0583fbb7dd833720a4 (current master) if I revert b4b34db65 I'm able to still reproduce the issue on my end. Curl still occasionally prints out:

```
$ curl --remote-name-all -Ss $(cat /urls.txt) -L -Z --parallel-max 100 --http2-prior-knowledge
curl: (55) Connection died, tried 5 times before giving up
```

**Comment by bagder** [comment]
How silly, so okey #7515 should fix that diversion...

**Comment by jay** [comment]
I'm able to reproduce this in Windows but rarely and I'm unable to get it down to fewer URLs to eliminate the noise. I had to put the URL list in a config style file (eg lines prefixed with url= like `url = http://theurl`) and passed them via config (curl ... -K urls.txt) due to shell differences. I looked through a wireshark [decrypted](https://everything.curl.dev/usingcurl/tls/sslkeylogfile) http2 capture and I see all redirects and nothing unusual. I should note I did it manually and may have missed something, since I don't have an automated way to scan the packet capture for unusual circumstance it's more of I know it when I see it.

**Comment by alexcrichton** [comment]
For your reproduction on Windows, were following redirects enabled? It's expected that the URL https://crates.io/api/v1/crates/quote/1.0.9/download redirects to https://static.crates.io/crates/quote/quote-1.0.9.crate, and at least when I was testing locally I had to ensure redirects were followed to get it to more reliably fail.

**Comment by jay** [comment]
Yes redirects were enabled but it occurs to me I had a capture filter on just crates.io and not any redirected to domains so I'll have to check again.

**Comment by Logarithmus** [comment]
@alexcrichton would applying https://github.com/curl/curl/commit/252790c5335a221179dfbacb40e065750bbe7544 as a reverse patch to `curl 7.78.0` fix the issue?

**Comment by alexcrichton** [comment]
To confirm, reverting both https://github.com/curl/curl/commit/b4b34db65f9f877384b00b9c20931bed5be1ff44 (to get nghttp2 detected locally for me) as well as https://github.com/curl/curl/commit/252790c5335a221179dfbacb40e065750bbe7544 appears to fix the issue on my end.

**Comment by bagder** [comment]
That commit fixed #6862 though, so presumable reverting this commit will bring back that back... Thoughts on this @ngg ?

**Comment by ngg** [comment]
There were 4 commits merged together in #6910 which solved #6862 (my original bug was that after a server returned HTTP_1_1_REQUIRED error on a HTTP/2 stream, curl would sometimes retry the connection over HTTP/2 again):

605e8423554e6c53c7755805f18354973b563b16 `http2: move the stream error field to the per-transfer storage`
I'm pretty sure this commit is necessary in some form, as it is much more logical to store stream errors per-transfer and not per-connection.

9c18c0b4dd8f00ed60f0c1994138afe556428114 `multi: don't close connection HTTP_1_1_REQUIRED`
This commit is necessary to not close other streams when one stream requires HTTP/1.1, I think it is ok as well.

9cb48457c6a9db533d6759a31a49ba13f1bc591b `ConnectionExists: respect requests for h1 connections better`
This commit is necessary as well to choose a HTTP/1.1 connection from the pool when it is explicitly needed (either by forcing HTTP/1.1 from curl_easy_setopt ot after downgrading the connection due to an error)

252790c5335a221179dfbacb40e065750bbe7544 `http2: call the handle-closed function correctly on closed stream`
This is the commit in question, and I don't really know what it actually does.
Its description says `This was this one condition where the stream could be closed due to an error and the function would still wrongly just return 0 for it.`
The comment in this branch `/* closed overrides paused */` is a bit strange because that case is seemingly already handled within the `if(httpc->pause_stream_id)` branch.
I don't think that reverting this change would change curl's behavior about HTTP_1_1_REQUIRED, but I don't really know what it actually does and what conditions are required to trigger that branch.

**Comment by bagder** [comment]
Thanks @ngg. I'm going to go ahead and merge the revert and we'll keep our eyes open and see if we run into any problems going forward.
